File: algorithmic_toolbox/week4_divide_and_conquer/4_number_of_inversions/python/inversions.py
import sys
from typing import MutableSequence


def get_number_of_inversions(a: MutableSequence[int], b: MutableSequence[int], left: int, right: int) -> int:
    number_of_inversions = 0

    if right <= left + 1:
        return number_of_inversions

    ave = (left + right) // 2
    number_of_inversions += get_number_of_inversions(a, b, left, ave)
    number_of_inversions += get_number_of_inversions(a, b, ave, right)
    number_of_inversions += merge(a, b, left, ave - 1, right - 1)

    return number_of_inversions


def merge(a: MutableSequence[int], b: MutableSequence[int], left: int, ave: int, right: int) -> int:  # noqa: C901
    i = left
    j = ave + 1
    k = left
    inv_count = 0

    while i <= ave and j <= right:
        if a[i] <= a[j]:
            b[k] = a[i]
            k += 1
            i += 1
        else:
            inv_count += ave - i + 1
            b[k] = a[j]
            k += 1
            j += 1

    while i <= ave:
        b[k] = a[i]
        k += 1
        i += 1

    while j <= right:
        b[k] = a[j]
        k += 1
        j += 1

    for i in range(left, right + 1):
        a[i] = b[i]

    return inv_count


if __name__ == "__main__":
    n, *a = list(map(int, sys.stdin.read().split()))
    b = n * [0]
    print(get_number_of_inversions(a, b, 0, len(a)))

File: algorithmic_toolbox/week4_divide_and_conquer/4_number_of_inversions/c/inversions.c
#include <stdio.h>
#include <stdlib.h>

long long merge(int a[], int b[], size_t left, size_t ave, size_t right) {
  size_t i = left;
  size_t j = ave + 1;
  size_t k = left;
  long long inv_count = 0;

  while (i <= ave && j <= right) {
    if (a[i] <= a[j]) {
      b[k] = a[i];
      k += 1;
      i += 1;
    } else {
      inv_count += ave - i + 1;
      b[k] = a[j];
      k += 1;
      j += 1;
    }
  }

  while (i <= ave) {
    b[k] = a[i];
    k += 1;
    i += 1;
  }

  while (j <= right) {
    b[k] = a[j];
    k += 1;
    j += 1;
  }

  for (i = left; i <= right; i++)
    a[i] = b[i];

  return inv_count;
}

long long getNumberOfInversions(int a[], int b[], size_t left, size_t right) {
  long long number_of_inversions = 0;

  if (right <= left + 1)
    return number_of_inversions;

  size_t ave = (left + right) / 2;
  number_of_inversions += getNumberOfInversions(a, b, left, ave);
  number_of_inversions += getNumberOfInversions(a, b, ave, right);
  number_of_inversions += merge(a, b, left, ave - 1, right - 1);

  return number_of_inversions;
}
